### PR TITLE
Opensea orders support

### PR DESCRIPTION
**Pull request by ex1st0r**

### CODE REVIEW

File: packages/protocol-ethereum-sdk/src/config/type.ts [hunk 1]
```diff
@@ -7,19 +8,73 @@ export type ExchangeFees = {
...
 	basePath: string
 	chainId: number
 	exchange: ExchangeAddresses
 	transferProxies: TransferProxies
+	proxyRegistries: ProxyRegistries
 	fees: ExchangeFees
 }
+
+
+export enum OrderOpenSeaV1DataV1FeeMethod {
```
Comment (evgenynacu) [inline packages/protocol-ethereum-sdk/src/config/type.ts:36]: this file is only for config related typed. OpenSea types should not be here

File: packages/protocol-ethereum-sdk/src/order/test/order.ts [hunk 2]
```diff
@@ -1,5 +1,5 @@
-import { toAddress, toBigNumber, toWord } from "@rarible/types"
-import { SimpleOrder } from "../sign-order"
+import {toAddress, toBigNumber, toWord} from "@rarible/types"
```
Comment (evgenynacu) [inline packages/protocol-ethereum-sdk/src/order/test/order.ts:1]: strange, I think this should not be possible when eslint config is applied

File: packages/protocol-ethereum-sdk/src/order/sign-order.ts [hunk 3]
```diff
@@ -41,12 +59,179 @@ export async function signOrder(
...
+		basePrice,
+		extra: toBigNumber(order.data.extra),
+		listingTime: toBigNumber(String(order.start || 0)),
+		expirationTime: toBigNumber(String(order.end || 0)),
+		salt: toBigNumber(toBn(order.salt).toString(10)),
+	}
+}
+
+export function hashOrder(order: OpenSeaOrderToSignDTO): string {
+	return Web3.utils.soliditySha3(
```
Comment (evgenynacu) [inline packages/protocol-ethereum-sdk/src/order/sign-order.ts:190]: let's use @ethersproject/abi. it's used in both web3.js and ethers.js

File: packages/protocol-ethereum-sdk/src/order/sign-order.ts [hunk 4]
```diff
@@ -41,12 +59,179 @@ export async function signOrder(
...
+		{type: "bytes32", value: hash}
+	) || ""
+}
+
+export function hashOpenSeaV1Order(ethereum: Ethereum, order: SimpleOpenSeaV1Order): string {
+	return hashOrder(convertOpenSeaOrderToSignDTO(ethereum, order))
+}
+
+//TODO replace web3.eth.sign
+export async function getOrderSignature(ethereum: Ethereum, order: SimpleOpenSeaV1Order): Promise<string> {
```
Comment (evgenynacu) [inline packages/protocol-ethereum-sdk/src/order/sign-order.ts:229]: should we move this to test only?

File: packages/protocol-ethereum-sdk/src/order/approve-opensea.ts [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+	const proxyAddress = await getRegisteredProxy(ethereum, config.proxyRegistries.openseaV1)
+
+	switch (asset.assetType.assetClass) {
+		case "ERC20": {
+			const contract = asset.assetType.contract
+			const operator = config.transferProxies.openseaV1
+			return approveErc20(ethereum, send, contract, owner, operator, asset.value, infinite)
+		}
+		case "ERC721": {
+			const contract = asset.assetType.contract
```
Comment (evgenynacu) [inline packages/protocol-ethereum-sdk/src/order/approve-opensea.ts:29]: where is the logic where new proxy is created for each user? didn't find it

**Review by evgenynacu — CHANGES_REQUESTED** [review]

File: packages/protocol-ethereum-sdk/src/order/sign-order.ts (hunk 3 above)
Comment (ex1st0r) [reply packages/protocol-ethereum-sdk/src/order/sign-order.ts]: done

File: packages/protocol-ethereum-sdk/src/order/sign-order.ts (hunk 4 above)
Comment (ex1st0r) [reply packages/protocol-ethereum-sdk/src/order/sign-order.ts]: done

File: packages/protocol-ethereum-sdk/src/order/approve-opensea.ts (hunk 5 above)
Comment (ex1st0r) [reply packages/protocol-ethereum-sdk/src/order/approve-opensea.ts:29]: You can find it in the current file, "getRegisteredProxy"

File: packages/protocol-ethereum-sdk/src/config/type.ts (hunk 1 above)
Comment (ex1st0r) [reply packages/protocol-ethereum-sdk/src/config/type.ts]: done

File: packages/protocol-ethereum-sdk/src/order/approve-opensea.ts [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+	ethereum: Ethereum,
+	proxyRegistry: Address
+): Promise<Address> {
+	const proxyRegistryContract = createOpenseaProxyRegistryEthContract(ethereum, proxyRegistry)
+	const from = toAddress(await ethereum.getFrom())
+	let proxyAddress = await getSenderProxy(proxyRegistryContract, from)
+
+	if (proxyAddress === ZERO_ADDRESS) {
+		await proxyRegistryContract.functionCall("registerProxy").send()
+		proxyAddress = await getSenderProxy(proxyRegistryContract, from)
```
Comment (kv9991) [inline packages/protocol-ethereum-sdk/src/order/approve-opensea.ts:54]: will it work as expected? I think it won't work in some cases. We should do polling for data after sending the transactions

Also we should support timeout for polling

Comment (evgenynacu) [reply packages/protocol-ethereum-sdk/src/order/approve-opensea.ts:54]: we should do tx.wait(). then it'll be there. also, it's good to do retry for this (if address is zero)

Comment (evgenynacu) [reply packages/protocol-ethereum-sdk/src/order/approve-opensea.ts:54]: this works on ganache because txs are confirmed immediately, but won't work on real networks

File: packages/protocol-ethereum-sdk/src/order/approve-opensea.ts [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+export async function approveOpensea(
+	ethereum: Ethereum,
+	send: SendFunction,
+	config: Config,
+	owner: Address,
+	asset: Asset,
+	infinite: undefined | boolean = true
+): Promise<EthereumTransaction | undefined> {
+
+	const proxyAddress = await getRegisteredProxy(ethereum, config.proxyRegistries.openseaV1)
```
Comment (evgenynacu) [inline packages/protocol-ethereum-sdk/src/order/approve-opensea.ts:20]: should we do this only when asset type is ERC721 or ERC1155?

File: packages/protocol-ethereum-sdk/src/order/approve-opensea.ts (hunk 6 above)
Comment (evgenynacu) [reply packages/protocol-ethereum-sdk/src/order/approve-opensea.ts:54]: also, this should throw if after getting the result, it's 0x0

File: packages/protocol-ethereum-sdk/src/order/fill-order.ts [hunk 8]
```diff
@@ -102,6 +137,131 @@ async function fillOrderV2(
...
+	buyer: Address
+) {
+	let buy: SimpleOpenSeaV1Order, sell: SimpleOpenSeaV1Order
+
+	const inverted = invertOrder(order, toBn(amount), buyer, order.salt)
+	const feeRecipient = order.data.feeRecipient === ZERO_ADDRESS ? buyer : ZERO_ADDRESS
+
+	switch (order.data.side) {
+		case "SELL": {
+			sell = {...order, taker: ZERO_ADDRESS}
```
Comment (evgenynacu) [inline packages/protocol-ethereum-sdk/src/order/fill-order.ts:153]: why do you set taker as ZERO? It'll be another order, so signature check can fail

File: packages/protocol-ethereum-sdk/src/order/fill-order.ts [hunk 9]
```diff
@@ -102,6 +137,131 @@ async function fillOrderV2(
...
+			buy = {
+				...inverted,
+				data: {...order.data, feeRecipient, side: "BUY"},
+			}
+			break
+		}
+		case "BUY": {
+			buy = {...order, taker: ZERO_ADDRESS}
```
Comment (evgenynacu) [inline packages/protocol-ethereum-sdk/src/order/fill-order.ts:161]: the same here with taker

File: packages/protocol-ethereum-sdk/src/order/fill-order.ts [hunk 10]
```diff
@@ -102,6 +137,131 @@ async function fillOrderV2(
 	return matchOrders(getMakeFee, ethereum, send, contract, request.order, orderRight)
 }
 
+export function getOpenseaOrdersForMatching(
```
Comment (evgenynacu) [inline packages/protocol-ethereum-sdk/src/order/fill-order.ts:140]: one small thing (and not only in this file) functions used by other functions usually go below (so you read what's more high-level first)

File: packages/protocol-ethereum-sdk/src/order/fill-order.ts [hunk 11]
```diff
@@ -102,6 +137,131 @@ async function fillOrderV2(
...
+	exchange: Address,
+	sell: SimpleOpenSeaV1Order,
+	buy: SimpleOpenSeaV1Order,
+) {
+	const sellOrderToSignDTO = convertOpenSeaOrderToSignDTO(ethereum, sell)
+	const buyOrderToSignDTO = convertOpenSeaOrderToSignDTO(ethereum, buy)
+
+	const exchangeContract = createOpenseaContract(ethereum, exchange)
+
+	const makerVRS = toVrs(sell.signature || buy.signature || "0x")
```
Comment (evgenynacu) [inline packages/protocol-ethereum-sdk/src/order/fill-order.ts:240]: that's strange logic, when you try to use the same signature for both orders. you should use 2 vrs values (sell and buy). One of them can be empty - if order.maker is executing transaction

File: packages/protocol-ethereum-sdk/src/order/fill-order.ts [hunk 12]
```diff
@@ -102,6 +137,131 @@ async function fillOrderV2(
...
+		sellOrderToSignDTO.calldata,
+		buyOrderToSignDTO.replacementPattern,
+		sellOrderToSignDTO.replacementPattern,
+		buyOrderToSignDTO.staticExtradata,
+		sellOrderToSignDTO.staticExtradata,
+		[makerVRS.v, makerVRS.v],
+		[makerVRS.r, makerVRS.s, makerVRS.r, makerVRS.s, "0x0000000000000000000000000000000000000000000000000000000000000000"],
+	)
+
+	// debugger
```
Comment (evgenynacu) [inline packages/protocol-ethereum-sdk/src/order/fill-order.ts:260]: unneeded comment

File: packages/protocol-ethereum-sdk/src/order/fill-order.ts [hunk 13]
```diff
@@ -102,6 +137,131 @@ async function fillOrderV2(
...
+export function getOpenseaOrdersForMatching(
+	ethereum: Ethereum,
+	order: SimpleOpenSeaV1Order,
+	amount: number,
+	buyer: Address
+) {
+	let buy: SimpleOpenSeaV1Order, sell: SimpleOpenSeaV1Order
+
+	const inverted = invertOrder(order, toBn(amount), buyer, order.salt)
+	const feeRecipient = order.data.feeRecipient === ZERO_ADDRESS ? buyer : ZERO_ADDRESS
```
Comment (evgenynacu) [inline packages/protocol-ethereum-sdk/src/order/fill-order.ts:149]: why feeRecipient is buyer? I think, this won't be the case (cause it will be always filled with opensea address), but we should set our address for this. We can set in in the config

**Comment by evgenynacu** [comment]
Also, check getBaseOrderFillFee. It should return what fee will be deducted from the user when order is filled.
And it's base fee (which is not controlled by this user). In this case, it should be calculated from the Order.

So, for example, if order is sell side then base fee (which can't be controlled by the buyer) is sellOrder.takerRelayerFee (as i see from opensea's contracts)

If this is BUY order, then it's the same, buyOrder.taker relayer fee = base fee
Pls, check this logic too. It's in executeFundsTransfer

So, basically, getBaseOrderFillFee should return what fee will be deducted from current user (which is filling the order)
